dbms/src/Columns/ColumnDecimal.cpp

```diff
@@ -140,59 +140,166 @@ const char * ColumnDecimal<T>::deserializeAndInsertFromArena(const char * pos, c
 }
 
 template <typename T>
-void ColumnDecimal<T>::countSerializeByteSizeForCmp(PaddedPODArray<size_t> & byte_size, const TiDB::TiDBCollatorPtr &)
-    const
+void ColumnDecimal<T>::countSerializeByteSizeForCmp(
+    PaddedPODArray<size_t> & byte_size,
+    const NullMap * nullmap,
+    const TiDB::TiDBCollatorPtr &) const
 {
-    countSerializeByteSizeImpl<true>(byte_size);
+    if (nullmap != nullptr)
+        countSerializeByteSizeImpl<true, true>(byte_size, nullmap);
+    else
+        countSerializeByteSizeImpl<true, false>(byte_size, nullptr);
 }
 
 template <typename T>
 void ColumnDecimal<T>::countSerializeByteSize(PaddedPODArray<size_t> & byte_size) const
 {
-    countSerializeByteSizeImpl<false>(byte_size);
+    countSerializeByteSizeImpl<false, false>(byte_size, nullptr);
 }
 
-
 template <typename T>
 void ColumnDecimal<T>::countSerializeByteSizeForCmpColumnArray(
     PaddedPODArray<size_t> & byte_size,
     const IColumn::Offsets & array_offsets,
+    const NullMap * nullmap,
     const TiDB::TiDBCollatorPtr &) const
 {
-    countSerializeByteSizeForColumnArrayImpl<true>(byte_size, array_offsets);
+    if (nullmap != nullptr)
+        countSerializeByteSizeForColumnArrayImpl<true, true>(byte_size, array_offsets, nullmap);
+    else
+        countSerializeByteSizeForColumnArrayImpl<true, false>(byte_size, array_offsets, nullptr);
 }
 
 template <typename T>
 void ColumnDecimal<T>::countSerializeByteSizeForColumnArray(
     PaddedPODArray<size_t> & byte_size,
     const IColumn::Offsets & array_offsets) const
 {
-    countSerializeByteSizeForColumnArrayImpl<false>(byte_size, array_offsets);
+    countSerializeByteSizeForColumnArrayImpl<false, false>(byte_size, array_offsets, nullptr);
+}
+
+template <typename T>
+template <bool compare_semantics, bool has_nullmap>
+void ColumnDecimal<T>::countSerializeByteSizeImpl(PaddedPODArray<size_t> & byte_size, const NullMap * nullmap) const
+{
+    RUNTIME_CHECK_MSG(byte_size.size() == size(), "size of byte_size({}) != column size({})", byte_size.size(), size());
+
+    size_t size = byte_size.size();
+    static constexpr T def_val{};
+    for (size_t i = 0; i < size; ++i)
+    {
+        if constexpr (compare_semantics && is_Decimal256)
```

Comment:
What if `compare_semantics` is true and `is_Decimal256` is false? It looks like the `nullmap` will be ignored.

Reply:
Still need to add `sizeof(T)` to `byte_size[i]` if row[i] is null. So skip check nullmap